### PR TITLE
Add %deprecate pragma

### PR DESCRIPTION
**Pull request by mattpolzin**
Add the `%deprecate` pragma as proposed [here](https://github.com/idris-lang/Idris2/issues/2082).

Closes https://github.com/idris-lang/Idris2/issues/2082.

In short, add `%deprecate` before a definition to cause the compiler to generate a warning whenever the given function is applied. Modify or add code documentation (`||| docs`) for the definition to give any additional suggestions you want printed along with the warning.

- Adds `%deprecate` pragma and associated warning.
- Updates CHANGELOG
- Adds beginnings of a "pragmas" page in the reference documentation.

### CODE REVIEW

File: src/Idris/Doc/String.idr [hunk 1]
```diff
@@ -454,7 +466,8 @@ getDocsForName fc n config
                   in annotate DocStringBody
                      (concatWith (\l, r => l <+> hardline <+> r) docs)
                      <$ guard (not (null docs))
-             pure (vcat (docDecl :: docBody))
+             let maybeDocDecl = [docDecl | showType]
+             pure . vcat . catMaybes $ maybeDocDecl :: (map Just $ docBody)
```
Comment (gallais) [inline src/Idris/Doc/String.idr:470]: Should the `:doc` output also include a `DEPRECATED` string?

Comment (mattpolzin) [reply src/Idris/Doc/String.idr:470]: Yeah, that seems smart. Will do.

**Review by gallais — COMMENTED** [review]
Does the fact `checkDeprecation` is called from `CompileExpr` mean that we get away with
using deprecated functions in values that have a `0` quantity and are not compiled?
Should this be checked during variable lookup in `TTImp.Elab.App` instead?

**Comment by mattpolzin** [comment]
> Does the fact `checkDeprecation` is called from `CompileExpr` mean that we get away with
> 
> using deprecated functions in values that have a `0` quantity and are not compiled?

Not sure, I'll try that and see. 

> Should this be checked during variable lookup in `TTImp.Elab.App` instead?

Quite possibly. Thanks for the idea. Where to check this was roughly a guess on my part so I only know that it works for the cases I tested but not that it's the best place to put it.

**Comment by mattpolzin** [comment]
Much better results; in broad strokes moving to `Elab.App` was the prefect plan, but let me know if the way I surfaced it there seems reasonable to you!

File: src/TTImp/Elab/App.idr [hunk 2]
```diff
@@ -91,17 +98,27 @@ getNameType rigc env fc x
...
                            (throw (LinearMisuse fc !(getFullName x) lhs rhs))
 
+    checkDeprecation : FC -> GlobalDef -> Core ()
+    checkDeprecation fc gdef =
+      do when (Deprecate `elem` gdef.flags) $
+           recordWarning $
+             Deprecated
+               "\{show gdef.fullname} is deprecated and will be removed in a future version."
+               (Just (fc, gdef.fullname))
+    checkDeprecation _ _ = pure ()
```
Comment (buzden) [inline src/TTImp/Elab/App.idr:108]: Isn't this line an unreachable case?

Comment (mattpolzin) [reply src/TTImp/Elab/App.idr]: It certainly is. Fixed!

**Review by gallais — APPROVED** [review]

**Pull request merged**